src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/ObjectWriter/R2RPEBuilder.cs

```diff
@@ -560,25 +557,32 @@ static class PEHeaderCopier
         /// <param name="target">Target architecture to set in the header</param>
         public static PEHeaderBuilder Copy(PEHeaders peHeaders, TargetDetails target)
         {
-            // Default base addresses used by Roslyn
-            const ulong DefaultExeBaseAddress64Bit = 0x1_4000_0000;
-            const ulong DefaultDllBaseAddress64Bit = 0x1_8000_0000;
+            return Create(
+                peHeaders.CoffHeader.Characteristics,
+                peHeaders.PEHeader.DllCharacteristics,
+                peHeaders.PEHeader.Subsystem,
+                target);
+        }
 
+        /// <summary>
+        /// Fill in PE header information into a PEHeaderBuilder used by PEBuilder.
+        /// </summary>
+        /// <param name="relocsStripped">Relocs are not present in the PE executable</param>
+        /// <param name="dllCharacteristics">Extra DLL characteristics to apply</param>
+        /// <param name="subsystem">Targeting subsystem</param>
+        /// <param name="target">Target architecture to set in the header</param>
+        public static PEHeaderBuilder Create(Characteristics characteristics, DllCharacteristics dllCharacteristics, Subsystem subsystem, TargetDetails target)
+        {
             bool is64BitTarget = target.PointerSize == sizeof(long);
 
-            Characteristics imageCharacteristics = peHeaders.CoffHeader.Characteristics;
-            if (is64BitTarget)
-            {
-                imageCharacteristics &= ~Characteristics.Bit32Machine;
-                imageCharacteristics |= Characteristics.LargeAddressAware;
-            }
+            characteristics |= (is64BitTarget ? Characteristics.LargeAddressAware : Characteristics.Bit32Machine);
```

Comment:
This isn't doing the same as the code it replaces. In particular, the old code when presented with a Bit32Machine entry was clearing it.

Reply:
I believe the proper way to set these bits is to have them mutually exclusive - i.e. in 32-bit mode, LargeAddressAware is cleared and Bit32Machine is set and vice versa in 64-bit mode. IIRC LargeAddressAware + Bit32Machine is the combo used for AnyCPU MSIL but that doesn't make sense for the output R2R native image. I have modified the change in this sense, thanks for pointing out this inconsistency!